### PR TITLE
fix(fws): Sound scheduling for FWS, triple click on capability downgrade

### LINKED ISSUE #7710: "autobrake off" callout
### Aircraft Version

Experimental

### Description

Include an "autobrake off" callout in the FBW A32NX. Reference below. It's a sound effect, triggers a master caution as well.

### References (optional)

https://youtu.be/oaIUNfbxYmM

### Additional info (optional)

If possible, make it selectable in the EFB whether to use it or not.

### Discord Username (optional)

DerbeQ#9424

### LINKED ISSUE #9008: [A380X] Better AP Disconnect Logic
### Aircraft Version

Development

### Build info

```json
{
    "built": "2024-09-30T16:42:48+00:00",
    "ref": "refs/heads/a380x-experimental",
    "sha": "c090953517e912d5833db4783ac671c61fb1db69",
    "actor": "aguther",
    "event_name": "manual",
    "pretty_release_name": "a380x-experimental:c0909535",
    "version": "a380x-v0.12.0-a380x-experimental.c090953"
}
```

### Describe the bug

Currently when disconnecting the autopilot, the events are as follow:
- Autopilot Disconnect
- 3 Autopilot disconnect chimes
- 3 Landing Capability Downgraded clicks

No matter the number of click on the sidestick takeover button, the sequence is the same

### Expected behavior

When disconnecting the autopilot, the sequence should be:
-  Autopilot Disconnect
- Master Warning Light should come on and stay on until the sidestick takeover has been clicked a second time or the master warning button has been pushed
- Autopilot disconnect chimes until the sidestick takeover has been clicked a second time or the master warning button has been pushed
- 3 Landing Capability Downgraded clicks if switching from an ILS CAT3 approach to an ILS CAT1 (iirc)

### Steps to reproduce

1. Fly above 400ft AGL
2. Engage autopilot
3. Disengage autopilot
4. Observe the events after autopilot is off

### References (optional)

https://www.youtube.com/watch?v=qcXQKQukxkI&t=59s

& many other A380 cockpit landing videos

### Additional info (optional)

_No response_

### Discord Username (optional)

PilotEyesA350

### PR DESCRIPTION
**Pull request by flogross89**
Fixes #9008
Fixes #7710 

## Summary of Changes

Rework of some FWS parts for A32NX and A380X concerning sound scheduling.

**A32NX + A380X:** 

- Adds proper triple click logic: Triple click is scheduled for playback if approach capability is downgraded. Around 1 seconds debouncing and 3 sec suppression after a downgrade, since our downgrade from CAT3D to CAT 1 takes quite a while

**A380X:** 

- Move all relevant sounds (ROP/ROW, autobrake off, AP/ATHR disc., triple click, altitude callouts) to FwsSoundManager

**A32NX:** 

- Add "autobrake off" callout
- Refine AP/ATHR disconnection logic
- Move some, but not all relevant sounds (autobrake off, AP/ATHR disc., triple click) to FwsSoundManager

## Additional context

Discord username (if different from GitHub): floridude

## Testing instructions

**A32NX:**
- Prepare normal A-B flight & take off
- Disconnect AP and A/THR multiple times during the flight
- a. By using a joystick and the AUTOPILOT/AUTOTHROTTLE OFF/DISCONNECT bindings. One press should trigger an AP OFF ECAM memo and an aural cavalry charge, a second press within 0.8s should silence it. If you wait longer than 0.8s with the second press, you should get a continuous cavalry charge, an ECAM memo and an ECAM fault "AP OFF"
- b. By clicking the AP push button on the FCU. This should trigger a continuous cavalry charge, an ECAM memo and an ECAM fault "AP OFF"
-->Triple clicks should not be heard when disconnecting the AP outside of an ILS approach
- Ensure that altitude c-chord (when leaving FCU altitude) works as expected
- Fly an ILS approach, engage both APs. Try various engage/disengage conditions. When the landing capability degrades, a triple click should be heard
- Land with engaged auto brake mode. Disconnect auto brake above 50kts, an "auto brake off" callout should be heard

**A380X:**
Try to trigger all of these callouts and confirm that they still sound as expected:
V1, autobrake off, AP disconnect, A/THR deactivation, altitude callouts, triple clicks, GPWS (terrain/pull up, ...)

Triple clicks shouldn't trigger for every AP disconnect, but only when the landing capability degrades. 
Ensure that altitude c-chord (when leaving FCU altitude) works as expected

## How to download the PR for QA

Every new commit to this PR will cause new A32NX and A380X artifacts to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, find and click on the **PR Build** tab
1. Click on either **flybywire-aircraft-a320-neo**, **flybywire-aircraft-a380-842 (4K)** or **flybywire-aircraft-a380-842 (8K)** download link at the bottom of the page

### CODE REVIEW

File: fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts [hunk 1]
```diff
@@ -1332,6 +1503,194 @@ export class PseudoFWC {
...
+        appCap = 0;
+      }
+      return appCap;
+    };
+
+    if (this.fmgc1DiscreteWord4.get().isNormalOperation()) {
+      fmgcApproachCapability = getApproachCapability(this.fmgc1DiscreteWord4.get());
+    } else if (this.fmgc2DiscreteWord4.get().isNormalOperation()) {
+      fmgcApproachCapability = getApproachCapability(this.fmgc1DiscreteWord4.get());
+    }
```
Comment (2hwk) [inline fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts:1596]:
```suggestion
    if (this.fmgc1DiscreteWord4.get().isNormalOperation() || this.fmgc2DiscreteWord4.get().isNormalOperation()) {
      fmgcApproachCapability = getApproachCapability(this.fmgc1DiscreteWord4.get());
    }
```

File: fbw-a380x/src/systems/systems-host/systems/FlightWarningSystem/FwsFlightPhases.ts [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+    this._updateTakeoffMemo(deltaTime);
+    this._updateLandingMemo(deltaTime);
+    this._updateAltitudeWarning();
+  }
+
+  _updateFlightPhase(_deltaTime: number) {
+    const raHeight1InValid = this.fws.radioHeight1.isFailureWarning() || this.fws.radioHeight1.isNoComputedData();
+    const raHeight2InValid = this.fws.radioHeight2.isFailureWarning() || this.fws.radioHeight2.isNoComputedData();
+    let radioHeight;
+    if (raHeight1InValid) {
```
Comment (2hwk) [inline fbw-a380x/src/systems/systems-host/systems/FlightWarningSystem/FwsFlightPhases.ts:164]: Is this meant to be invalid or in valid? I'm assuming its invalid so maybe the var naming should be `raHeight1Invalid` and `raHeight2Invalid`

File: fbw-a380x/src/systems/systems-host/systems/FlightWarningSystem/FwsFlightPhases.ts [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+    const ias = this.fws.computedAirSpeedToNearest2.get();
+    const acSpeedAbove80kts = this.speedAbove80KtsMemo.write(ias > 83, ias < 77);
+
+    const v1 = SimVar.GetSimVarValue('L:AIRLINER_V1_SPEED', 'knots');
+    let acAboveV1: boolean;
+    if (v1) {
+      acAboveV1 = this.speedAboveV1Memo.write(ias > v1 + 3, ias < v1 - 3);
+    } else {
+      acAboveV1 = false;
+    }
```
Comment (2hwk) [inline fbw-a380x/src/systems/systems-host/systems/FlightWarningSystem/FwsFlightPhases.ts:200]:
```suggestion
    acAboveV1 = v1 ? this.speedAboveV1Memo.write(ias > v1 + 3, ias < v1 - 3) : false;
```

File: fbw-a380x/src/systems/systems-host/systems/FlightWarningSystem/FwsFlightPhases.ts [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+      this._wasBellowThreshold = false;
+    }
+    if (delta >= 200 && delta <= 750) {
+      this._wasInRange = true;
+    }
+
+    if (this._wasBellowThreshold && this._wasReach200ft) {
+      if (delta >= 200) {
+        this.fws.soundManager.enqueueSound('cChordCont');
+      } else if (delta < 200) {
```
Comment (2hwk) [inline fbw-a380x/src/systems/systems-host/systems/FlightWarningSystem/FwsFlightPhases.ts:490]:
```suggestion
      } else {
```

File: fbw-a380x/src/systems/systems-host/systems/FlightWarningSystem/FwsFlightPhases.ts [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+      ) {
+        this.fws.soundManager.dequeueSound('cChordCont');
+        this.fws.soundManager.enqueueSound('cChordOnce');
+      }
+    } else if (delta > 750 && this._wasInRange && !this._wasReach200ft) {
+      if (delta > 750) {
+        this.fws.soundManager.enqueueSound('cChordCont');
+      } else if (delta >= 750) {
+        this.fws.soundManager.dequeueSound('cChordCont');
+      }
```
Comment (2hwk) [inline fbw-a380x/src/systems/systems-host/systems/FlightWarningSystem/FwsFlightPhases.ts:506]: (delta > 750) and (delta >= 750) feels like there's a mistake in the signage here, the 2nd condition seems like dead code here?

**Review by 2hwk — DISMISSED** [review]

File: fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts (hunk 1 above)
Comment (flogross89) [reply fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts:1596]: That's a typo in the original code, it's meant to take either Fmgc1 or Fmgc2 discreteword4. Latest commit fixes this

File: fbw-a380x/src/systems/systems-host/systems/FlightWarningSystem/FwsFlightPhases.ts (hunk 2 above)
Comment (flogross89) [reply fbw-a380x/src/systems/systems-host/systems/FlightWarningSystem/FwsFlightPhases.ts:164]: invalid, but this code was copied from the previous flight phase implementation

File: fbw-a380x/src/systems/systems-host/systems/FlightWarningSystem/FwsFlightPhases.ts (hunk 3 above)
Comment (flogross89) [reply fbw-a380x/src/systems/systems-host/systems/FlightWarningSystem/FwsFlightPhases.ts:200]: while this improves the code, I'd not touch it right now. It's a 1:1 copy from the old code, and worked there

**Comment by flogross89** [comment]
@2hwk Thanks for the review, just pushed some changes and left comments

File: fbw-a380x/src/systems/systems-host/systems/FlightWarningSystem/FwsFlightPhases.ts (hunk 5 above)
Comment (flogross89) [reply fbw-a380x/src/systems/systems-host/systems/FlightWarningSystem/FwsFlightPhases.ts]: fixed (hopefully, I hope that's how it was intended)

File: fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts [hunk 6]
```diff
@@ -1332,6 +1503,194 @@ export class PseudoFWC {
...
+      this.requestMasterWarningFromApOff = false;
+      this.soundManager.dequeueSound('cavalryChargeOnce');
+      this.soundManager.dequeueSound('cavalryChargeCont');
+      this.soundManager.setVolume(FwsAuralVolume.Full);
+    }
+
+    this.autoPilotInstinctiveDiscPressedPulse.write(false, deltaTime);
+
+    // approach capability downgrade. Debounce first, then suppress for a certain amount of time
+    // (to avoid multiple triple clicks, and a delay which is too long)
```
Comment (BravoMike99) [inline fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts:1569]: On the real plane you can have multiple triple clicks for e.g., when disconnecting both autopilots during approach sometimes you get the cavalry charge followed by two triple clicks in a row, one due to the downgrade from CAT 3 Dual to CAT 3 single and another from CAT 3 single to CAT 1 ( https://youtu.be/pZAI18IRSnE?t=212). On the A380 it jumps straight away from CAT 3 to CAT 1 so it never happens it seems.

File: fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts [hunk 7]
```diff
@@ -1332,6 +1503,194 @@ export class PseudoFWC {
...
+      !apEngaged && !this.autoPilotOffVoluntaryMemory.read() && this.autoPilotOffUnacknowledged.read(),
+      !this.autoPilotOffUnacknowledged.read(),
+    );
+    this.autoPilotOffInvoluntary.set(this.autoPilotOffInvoluntaryMemory.read());
+    this.autoPilotOffShowMemo.set(this.autoPilotOffVoluntaryMemory.read() || this.autoPilotOffInvoluntaryMemory.read());
+
+    if (this.autoPilotDisengagedInstantPulse.read()) {
+      // Request quiet CRC one time
+      this.requestMasterWarningFromApOff = true;
+      this.soundManager.setVolume(FwsAuralVolume.Attenuated);
```
Comment (BravoMike99) [inline fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts:1552]: It should always be full volume, the only difference on the A320 for FWS volume is engines running vs engines off.  The minimum cavalry charges played are also quite different but that can be left for future work I guess as its out of scope of the intent of the PR  😄  .

**Review by BravoMike99 — COMMENTED** [review]
Just some small FWS logic comments due to differences vs the A380.

File: fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts (hunk 7 above)
Comment (flogross89) [reply fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts:1552]: Interesting! Then I'll remove that part. I believe this setting is without effect anyways at the moment. About the minimum number of cav charges: I implemented that differently for a32nx and a380x: a380x CC can be cancelled instantly, a32nx plays a sequence of three CCs I believe (That's the "cavalryChargeOnce" event, one sound file with a number of CCs). Have you tried it in the sim?

File: fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts (hunk 6 above)
Comment (flogross89) [reply fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts:1569]: Are these multiple triple clicks systematic? I believe there's a suppression of around 3 seconds if multiple downgrades happen. Might be some weird timing behavior which we see in the video. https://youtu.be/uPxlp83krBw?si=YfDG7blRXWTWii6C This one and many others have one triple click for a downgrade from C3D to C1

File: fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts [hunk 8]
```diff
@@ -2474,56 +2844,105 @@ export class PseudoFWC {
...
+    this.soundManager.handleSoundCondition(
+      'keepMaxReverse',
       ias <= 80 && ias > 4 && (w.bitValueOr(12, false) || w.bitValueOr(13, false)),
     );
   }
 
+  autoThrottleInstinctiveDisconnect() {
+    // When instinctive A/THR disc. p/b is pressed after ABRK deactivation, inhibit audio+memo, don't request master caution
+    // Unclear refs, whether this has to happen within the audio confirm node time (1s)
+    if (this.autoBrakeDeactivatedNode.read()) {
```
[the remark this answers is not in the document]
Comment (flogross89) [reply fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts:2900]: Good to know, thanks! Removed that part.

File: fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts [hunk 9]
```diff
@@ -2474,56 +2844,105 @@ export class PseudoFWC {
...
+      whichCodeToReturn: () => [0],
+      codesToReturn: ['220800001'],
+      memoInhibit: () => false,
+      failure: 3,
+      sysPage: -1,
+      side: 'LEFT',
+    },
+    220800004: {
+      // A/THR OFF involuntary
+      flightPhaseInhib: [3, 4, 5, 10],
```
[the remark this answers is not in the document]
Comment (flogross89) [reply fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts:2937]: Indeed they were not, oversight on my side. Fixed, thanks!

File: fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts [hunk 10]
```diff
@@ -4282,6 +4701,18 @@ export class PseudoFWC {
       sysPage: -1,
       side: 'RIGHT',
     },
+    // 32 LANDING GEAR
+    320000001: {
+      // AUTO BRK OFF
+      flightPhaseInhib: [1, 2, 3, 4, 5, 6, 7, 8, 9, 12],
```
[the remark this answers is not in the document]
Comment (flogross89) [reply fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts:4707]: also fixed

**Review by BravoMike99 — APPROVED** [review]

**Comment by BravoMike99** [comment]
Quality Assurance Tester

Discord Username  : bruno_pt99 
Object of testing : #9555 
Aircraft          : A32NX 
Tier of Testing   : 1
Date              : 23/11/2024 

Testing Process: 
 
Performed a normal flight and binded the following keys "AUTOPILOT OFF" & "AUTOTHROTTLE DISCONNECT"
1) Disconnected AP with double press in the air, 3 cavalry charges sounded and the "AP OFF" memo appeared alongside no triple click ✔️ .
2) Involuntarily disconnected the AP, AUTOFLT AP OFF appeared on the ECAM alongside "AP OFF" and the cavalry charge continued to sound till pressing the disconnect button ✔️ .
3) Disconnected ATHR via the bind, AUTO FLT ATHR OFF caution appeared instead of "ATHR OFF" memo ❌ 
4) Tested multiple mode reversions and triple click worked ✔️ 
5) Left the assigned FCU altitude in level flight and the Chord sounded ✔️ 
6) During approach. Did mode reversion from CAT 3D -> CAT 3 Single, CAT 3D/3S -> CAT 2 and triple click sounded ✔️ 
7) After disconnecting autopilot the triple click did not seem to sound after the downgrade to CAT 1, or, it is overlapping with the cavalry charges instead of playing after them ❌ 
8) During the flare, setting the thrust to idle triggered the AUTO FLT ATHR OFF ecam caution.❌ 
8) Autobrake OFF callout sounded above 50kt when disconnecting autobrake ✔️ 

Testing Results:  Not Passed  

Negatives:       

 I also noticed that sometimes when going idle thrust with my hardware levers it would disconnect with the "ATHR OFF" memo while other times with the AUTOFLT ATHR off (perhaps related to hardware noise?).  In addition, when the "ATHR OFF" memo was triggered, a single chime didn't sound, which it should.
Bonus: With the involuntary AP disconnection the ecam warning should remain there even after the takeover/disconnect button is pressed, the only way to get rid of it is via the CLR button. Currently pressing the disconnect button clears the warning.
Setting thrust to MCT or TOGA in the air trigerred the AUTOFLT ATHR off (see media)

Media:

MCT

File: fbw-a32nx/src/wasm/fbw_a320/src/interface/SimConnectInterface.cpp [hunk 11]
```diff
@@ -2431,6 +2431,9 @@ void SimConnectInterface::processEvent(const DWORD eventId, const DWORD data0, c
       simInputThrottles.ATHR_push = 1;
       std::cout << "WASM: event triggered: AUTO_THROTTLE_ARM" << std::endl;
       break;
+
+      // Re emitting masked event
+      execute_calculator_code("(>K:A32NX.AUTO_THROTTLE_DISCONNECT)", nullptr, nullptr, nullptr);
```
Comment (BravoMike99) [inline fbw-a32nx/src/wasm/fbw_a320/src/interface/SimConnectInterface.cpp:2436]: There's the culprit for the instinctive not working with the bind. Should be on the AUTO_THROTTLE_DISCONNECT case 😄

Comment (flogross89) [reply fbw-a32nx/src/wasm/fbw_a320/src/interface/SimConnectInterface.cpp:2436]: Ahhh damnit :D Thanks for finding that

**Comment by flogross89** [comment]
@BravoMike99 Thanks! Just pushed fixes for:
- single chime missing
- instinctive disconnect not recognized (involuntary triggered instead of voluntary)
Could you try the failed ones again? 

Can't explain 7. and 8. at the moment, it has the same logic as the a380x (which should work from what I know)

File: fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts [hunk 12]
```diff
@@ -1332,6 +1501,184 @@ export class PseudoFWC {
...
+      const land3FailPassiveCapacity = dw4.bitValueOr(24, false);
+      const land3FailOperationalCapacity = dw4.bitValueOr(25, false);
+
+      if (land2Capacity) {
+        appCap = 2;
+      } else if (land3FailPassiveCapacity) {
+        appCap = 3;
+      } else if (land3FailOperationalCapacity) {
+        appCap = 4;
+      } else if (landModeActive) {
```
Comment (BravoMike99) [inline fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts:1579]: The check for land mode armed is  missing which could explain why the triple click doesn't play with approach mode armed.

File: fbw-a32nx/src/wasm/fbw_a320/src/interface/SimConnectInterface.cpp [hunk 13]
```diff
@@ -1810,6 +1810,9 @@ void SimConnectInterface::processEvent(const DWORD eventId, const DWORD data0, c
       if (static_cast<long>(data0) == 1) {
         simInputAutopilot.AP_disconnect = 1;
         std::cout << "WASM: event triggered: AUTOPILOT_DISENGAGE_SET" << std::endl;
+
+        // Re emitting masked event for autopilot disconnection
+        execute_calculator_code("(>K:A32NX.AUTOPILOT_DISENGAGE)", nullptr, nullptr, nullptr);
```
[the remark this answers is not in the document]
Comment (flogross89) [reply fbw-a32nx/src/wasm/fbw_a320/src/interface/SimConnectInterface.cpp:1815]: How can I trigger that event again? Haven't seen anything about priorities.

File: fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts (hunk 12 above)
Comment (flogross89) [reply fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts]: added now, thanks for looking into it!

File: fbw-a32nx/src/wasm/fbw_a320/src/interface/SimConnectInterface.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (lukecologne) [reply fbw-a32nx/src/wasm/fbw_a320/src/interface/SimConnectInterface.cpp:1815]: Use `SimConnectInterface::sendEvent()` (or sendEventEx1), example usage from my FCU shim code in FlyByWireInterface.cpp:

```
simConnectInterface.sendEventEx1(SimConnectInterface::Events::HEADING_BUG_SET, SIMCONNECT_GROUP_PRIORITY_STANDARD,
                                   discreteOutputs.afs_outputs.hdg_trk_value, 1);
```

[the remark this answers is not in the document]
Comment (lukecologne) [reply fbw-a32nx/src/wasm/fbw_a320/src/interface/SimConnectInterface.cpp:1815]: And about priorities: https://docs.flightsimulator.com/html/Programming_Tools/SimConnect/SimConnect_API_Reference.htm#simconnect-priorities

[the remark this answers is not in the document]
Comment (flogross89) [reply fbw-a32nx/src/wasm/fbw_a320/src/interface/SimConnectInterface.cpp:1815]: But isn't it masked intentionally, to prohibit MSFS itself from getting the event? Not sure how priorities would help here. I just want to signal to the FWS that the AP has been disconnected. Don't know why the event was masked in the first place.

[the remark this answers is not in the document]
Comment (lukecologne) [reply fbw-a32nx/src/wasm/fbw_a320/src/interface/SimConnectInterface.cpp:1815]: Good point, not sure why it was masked originally, we used the custom AP for ages. It probably helps to just keep the sim's AP var always at false, to not toggle it and possibly show random data that does not actually reflect the state of our AP. If we want to keep it that way, a new custom event is definitely required. I would probably still add it to the simconnect events, and trigger it that way, much cleaner than using `execute_calculator_code`, at least IMO

[the remark this answers is not in the document]
Comment (flogross89) [reply fbw-a32nx/src/wasm/fbw_a320/src/interface/SimConnectInterface.cpp]: Can't still explain why the other AP events are not masked, while this one isn't :D Anyways, I replaced these calls with sendEvent() from SimConnect

File: fbw-a32nx/src/wasm/fbw_a320/src/interface/SimConnectInterface.cpp [hunk 14]
```diff
@@ -2436,6 +2440,9 @@ void SimConnectInterface::processEvent(const DWORD eventId, const DWORD data0, c
     case Events::AUTO_THROTTLE_DISCONNECT: {
       simInputThrottles.ATHR_disconnect = 1;
       std::cout << "WASM: event triggered: AUTO_THROTTLE_DISCONNECT" << std::endl;
+
+      // Re emitting masked event
+      execute_calculator_code("(>K:A32NX.AUTO_THROTTLE_DISCONNECT)", nullptr, nullptr, nullptr);
```
[the remark this answers is not in the document]
Comment (flogross89) [reply fbw-a32nx/src/wasm/fbw_a320/src/interface/SimConnectInterface.cpp]: Thanks, changed that also

**Comment by flogross89** [comment]
Reminder: #9615

File: fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts [hunk 15]
```diff
@@ -1332,6 +1507,194 @@ export class PseudoFWC {
...
+    this.approachCapabilityDowngradeDebouncePulse.write(this.approachCapabilityDowngradeDebounce.read(), deltaTime);
+    this.approachCapabilityDowngradeSuppress.write(this.approachCapabilityDowngradeDebouncePulse.read(), deltaTime);
+    // Capability downgrade after debounce --> triple click
+    if (this.approachCapabilityDowngradeDebouncePulse.read()) {
+      this.soundManager.enqueueSound('tripleClick');
+    }
+    this.fmgcApproachCapability.set(fmgcApproachCapability);
+
+    // A/THR OFF
+    const aThrEngaged = this.autoThrustStatus.get() === 2 || this.autoThrustMode.get() !== 0;
```
Comment (BravoMike99) [inline fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts:1624]: I believe this is the cause of the autothrust off message triggering when going MCT or TOGA when previously engaged, as the status will go to 1 (engaged_armed). In addition to that, is the mode check also required?

**Comment by BravoMike99** [comment]
> @BravoMike99 Thanks! Just pushed fixes for:
> 
> * single chime missing
> * instinctive disconnect not recognized (involuntary triggered instead of voluntary)
>   Could you try the failed ones again?
> 
> Can't explain 7. and 8. at the moment, it has the same logic as the a380x (which should work from what I know)

Just tried it again. Instinctive disconnect, single chime missing, triple click overlap & warning when going to MCT/TOGA are fixed now👍

**Comment by BravoMike99** [comment]
Quality Assurance Tester/Trainee Report

Discord Username  : bruno_pt99
Object of testing : #9555 
Aircraft          : A32NX 
Tier of Testing   : 2
Date              : 15/12/2024 

Testing Process: 
Performed a full flight and tested the following:
1) Disconnecting ATHR through instinctive and non instinctive methods and checked ATHR OFF memo & AUTOFLT AUTO THRUST OFF appeared accordingly ✔️ 
2) Disconnected AP voluntarily. AP OFF appeared on memo and by doing a second press it disappeared. If second is not pressed quickly enough, the AUTOFLT AP OFF warning appears with a continuous cavalry charge ✔️ 
3) Disconnecting AP off through non voluntary means also generated a continuous cavalry charge with the AUTOFLT AP OFF ecam warning and master warning light.
4) Altitude C chord sounded when leaving the assigned FCU altitude and by pressing the master warning button it silenced ✔️ 
5) Tested some FG mode reversions and the triple click worked ✔️ 
6) Tested multiple approach capability downgrades and the triple click sounded. Disconnecting autopilot without an approach capability downgrade did not trigger a triple click.   
7) Autobrake off callout sounded when disconnecting autobrake above 50kts. ✔️ 

Testing Results: 
Passed

Media:

https://github.com/user-attachments/assets/a36774fe-3723-41b7-b8ae-8034675a09fc

https://github.com/user-attachments/assets/46d53d90-7d0f-4b34-9dae-94a204b15ac0

https://github.com/user-attachments/assets/a5eaf646-13ff-4cf1-b107-f345d2dc1bb3


https://github.com/user-attachments/assets/47ae61e3-2917-4b74-b294-6f403503a82f

**Comment by BravoMike99** [comment]
Quality Assurance Tester

Discord Username  : bruno_pt99
Object of testing : #9555 
Aircraft          :  A380X 
Tier of Testing   : 1
Date              : 18/12/2024 

Testing Process: 
1) Tested multiple audio warning conditions (V1, AUTOBRAKE OFF, GPWS, RA callouts, master caution, master warning ) and they all sounded ✔️ 
2) Triple click did not sound when disconnecting autopilot without capability downgrade ✔️ 
3) Triple click sounded on mode reversion & approach capability downgrade.
4) Altitude horn sounded when leaving and reaching assigned FCU altitude ✔️ 

Testing Results: 
Passed

File: fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts [hunk 16]
```diff
@@ -70,20 +77,28 @@ enum FwcAuralWarning {
...
 
 export class PseudoFWC {
-  private readonly sub = this.bus.getSubscriber<A32NXFcuBusEvents & EwdSimvars & StallWarningEvents>();
+  private readonly sub = this.bus.getSubscriber<
+    PseudoFwcSimvars & StallWarningEvents & KeyEvents & A32NXFcuBusEvents
+  >();
+
+  private readonly fwsUpdateThrottler = new UpdateThrottler(125); // has to be > 100 due to pulse nodes
+
+  private keyEventManager: KeyEventManager;
```
Comment (tracernz) [inline fbw-a32nx/src/systems/systems-host/systems/FWC/PseudoFWC.ts:90]: Should be `private keyEventManager?: KeyEventManager;` as it is undefined after construction.

**Pull request merged**